spec/java/util/ArrayList.lsl

```diff
@@ -0,0 +1,1205 @@
+        action LIST_REMOVE(storage, index, 1);
+        modCount = modCount + 1;
+        length = length - 1;
+    }
+
+
+    proc _addElement (index: int, e: Object): void
+    {
+        _rangeCheckForAdd(index, length);
+        modCount = modCount + 1;
+        action LIST_INSERT_AT(storage, index, e);
+        length = length + 1;
+    }
+
+
+    proc _setElement (index: int, e: Object): Object
+    {
+        _checkValidIndex(index, length);
+        result = action LIST_GET(storage, index);
+        action LIST_SET(storage, index, element);
+    }
+
+
+    //methods
+
+    fun trimToSize (): void
+    {
+        modCount = modCount + 1;
+        //As we work with List we mustn't resize it... Or not ?
+    }
+
+
+    fun ensureCapacity (minCapacity: int): void
+    {
+        modCount = modCount + 1;
+        //As we work with List we mustn't resize it... Or not ?
+    }
+
+
+    fun size (): int
+    {
+        result = length;
+    }
+
+
+    fun isEmpty (): boolean
+    {
+        result = length == 0;
+    }
+
+
+    fun contains (o: Object): boolean
+    {
+        result = action LIST_FIND(storage, o, 0, length, +1) >= 0;
+    }
+
+
+    fun indexOf (o: Object): int
+    {
+        result = action LIST_FIND(storage, o, 0, length, +1);
+    }
+
+
+
+    fun lastIndexOf (o: Object): int
+    {
+        result = action LIST_FIND(storage, o, 0, length, -1);
+    }
+
+
+    fun clone (): Object
+    {
+        storageCopy = action LIST_DUP(storage);
+        result = new ArrayList(
+            state=self.state, storage=storageCopy, length=self.length);
+    }
+
+
+    fun toArray (): array<Object>
+    {
+        // #problem
+        //How to create a new array correctly ?
+        var a: array<int>;
+        result = action LIST_TO_ARRAY(storage, a, 0, length);
+    }
+
+
+    fun toArray (a: list<Object>): array<Object>
+    {
+        result = action LIST_TO_ARRAY(storage, a, 0, length);
+    }
+
+
+    fun get (index: int): Object
+    {
+        _checkValidIndex(index, length);
+        result = action LIST_GET(storage, index);
+    }
+
+
+    fun set (index: int, element: Object): Object
+    {
+        result = _setElement(index, element);
+    }
+
+
+    fun add (e: Object): boolean
+    {
+        modCount = modCount + 1;
+        action LIST_INSERT_AT(storage, length, e);
+        length = length + 1;
+        result = true;
+    }
+
+
+    fun add (index: int, element: Object): void
+    {
+        _addElement(index, e);
+    }
+
+
+    fun remove (index: int): Object
+    {
+        result = _deleteElement(index);
+    }
+
+
+    fun equals (other: Object): boolean
+    {
+        if (other == self)
+        {
+            result = true;
+        }
+        else
+        {
+            // #problem
+            //We don't know at this moment how create this method.
+            action NOT_IMPLEMENTED();
+        }
+    }
+
+
+    fun hashCode (): int
+    {
+        // result = action OBJECT_HASH_CODE(self);
```

Comment:
I don't see a problem here as we have the `OBJECT_HASH_CODE` action.
I think it would be just `hash(storage)`, doesn't it?

Reply:
@dog-m I agree with you; Was fixed;